### PR TITLE
Cylc registration fix.

### LINKED ISSUE #2749: cylc run <abs-src-dir> - bug or feature?
Standard/intended way to run a suite:
```
$ cylc register <NAME> <PATH>
$ cylc run <NAME>
# => suite run directory is $HOME/cylc-run/<NAME>/
$ cylc ping <NAME>  # etc.
```
@kinow's experiments with Docker have revealed that we can do this:
```
$ cylc run <PATH>
# => suite run directory is <PATH>
$ cylc ping <PATH>  # etc.
```
where PATH is the absolute path of the suite source directory.  Everything seems to work OK, except (e.g.) that `cylc scan` has to be forced to port scan because the filesystem scan under `$HOME/cylc-run` won't find the suite. The abs path works as the suite name for targeting the server.

@matthewrmshin - this strikes me as somewhat surprising ... is it known/intended behaviour that we should document, or an accident that we should prevent from working?

### LINKED ISSUE #2758: registration command line arg order
Correct arg order is `cylc reg <NAME> <PATH>` 

(Note that suite names can be hierarchical with a `/` delimiter, which can make it difficult to determine if the arg is meant to be a name or a directory path!).

Accidental arg reversal `cylc reg <PATH> <NAME>` causes problems:
 1. if `<NAME>` happens to be an existing suite source directory, this will register that suite with the name `<PATH>` (with run dir `~/cylc-run/<PATH>`)
 2. otherwise, if `<NAME>` is not an existing suite source directory, registration will appear to succeed, but the suite service directory will be created inside the source directory and nothing happens under `~/cylc-run`.

### PR DESCRIPTION
**Pull request by hjoliver**
Close #2758 
Close #2749 

[UPDATED description 15 Oct]:

Abort suite registration if:
* `suite.rc`not found in target directory (this allows us to distinguish correct arg order)
* name is an absolute path (can only be relative).
* name is already registered to another suite
  - but allow forced re-use of existing name and run directory with `--redirect`

Also support:
- auto-registration of suite in `$PWD`
  - `cylc reg NAME` (auto-reg)
  - `cylc reg` (auto reg and auto-name - with name of parent dir)
- on-the-fly  auto-registration by "cylc run" (of suite in `$PWD`, with name of parent dir).
  - replacing (via `exec`) process `cylc run` with `cylc run NAME` for easy identification with `ps` etc.

### CODE REVIEW

File: bin/cylc-register [hunk 1]
```diff
@@ -18,36 +18,37 @@
 
 """cylc [prep] register [OPTIONS] ARGS
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply bin/cylc-register]:
```
➤ cylc reg --help | head -n 1                                                         │
Usage: cylc [prep] register [OPTIONS] REG [PATH]
```
Like magic! :grin:

[the remark this answers is not in the document]
Comment (oliver-sanders) [reply bin/cylc-register]: Ah, forgot about that little bit of sorcery .

[the remark this answers is not in the document]
Comment (hjoliver) [reply bin/cylc-register]: ("ARGS" gets dynamically replaced with actual args, command-specific)

**Comment by hjoliver** [comment]
Travis CI is maddening at the moment, but I see one genuine test fail due to relying on the old reg behaviour...

File: lib/cylc/suite_srv_files_mgr.py [hunk 2]
```diff
@@ -422,24 +422,26 @@ def register(self, reg, source=None):
...
             if not os.path.isabs(source):
                 # On AIX on GPFS os.path.abspath(source) returns the source
                 # with full 'fileset' prefix. Manual use of $PWD to absolutize
                 # a relative path gives a cleaner result.
                 source = os.path.join(os.getenv("PWD", os.getcwd()), source)
-            source = os.path.normpath(source)
-            if (os.path.abspath(source) ==
-                    os.path.abspath(os.path.dirname(srv_d))):
-                source_str = ".."
-            else:
```
Comment (hjoliver) [inline lib/cylc/suite_srv_files_mgr.py]: Not sure what 436-8 was doing?

Comment (matthewrmshin) [reply lib/cylc/suite_srv_files_mgr.py]: Likely something we forgot to remove.

**Comment by oliver-sanders** [comment]
```console
$ ls
suite.rc
$ cylc reg $PWD
ERROR: no suite.rc found in /var/tmp/tmp.tHegPaMTo9
```

**Comment by hjoliver** [comment]
@oliver-sanders - technically that's user error. `cylc register NAME [PATH]` - name is not optional, but path is.  But fair enough, I should handle that error better.  I suppose we can say a hierarchical suite name can look like a relative file path, but not an absolute path, for one check...

**Comment by hjoliver** [comment]
I'll abort with an error if there is only one argument and it is either an absolute path or a valid relative path from cwd that is not also an existing registered suite name.  Does that sounds reasonable?

**Comment by oliver-sanders** [comment]
That sounds a bit complicated, could we just reject any `NAME` which is an absolute path?

From the help page I would have expected `cylc reg $PWD` to be equivalent to `cylc reg $PWD .`:

```console
$ cylc reg $PWD
ERROR: no suite.rc found in /var/tmp/tmp.0Rbo39DoUR
$ cylc reg $PWD .
REGISTER /var/tmp/tmp.0Rbo39DoUR: /var/tmp/tmp.0Rbo39DoUR
```

The leading slash messes things up:

```bash
# register suite in ~/cylc/run/var/tmp/tmp.0Rbo39DoUR
cylc reg var/tmp/tmp.0Rbo39DoUR /var/tmp/tmp.0Rbo39DoUR

# register suite in /var/tmp/tmp.0Rbo39DoUR
cylc reg /var/tmp/tmp.0Rbo39DoUR /var/tmp/tmp.0Rbo39DoUR
```

Do we currently make use of this behaviour?

**Comment by hjoliver** [comment]
> From the help page I would have expected `cylc reg $PWD` to be equivalent to `cylc reg $PWD .`

Que?  The help page (on this branch) only says you can omit PATH (2nd arg), not NAME, so neither of those examples should be allowed.  We definitely shouldn't be able to use a leading / in a NAME.

The complication is just that a relative path can't be distinguished from a name, which makes it difficult to know if the user got the args reversed, or if he/she omitted one, was it the right one...

**Comment by hjoliver** [comment]
(Sorry, just re-read above, I guess you intended `$PWD` as a NAME, in those examples?)

**Comment by oliver-sanders** [comment]
Indeed.

**Comment by hjoliver** [comment]
OK, let's make it that a NAME can be the same as the relative source path, because that actually seems like a sensible thing to do, but it can't be the same as the absolute source path, because the name gets used as a relative path under ~/cylc-run.

**Comment by hjoliver** [comment]
It might be nice to be able to `cylc reg PATH` as a shortcut for `cylc reg $(basename PATH) PATH`, or even `cylc reg` (no args) as a shortcut for the same, with `$PWD` as `PATH`.  Or is this getting more than the legal amount of complicated...

**Comment by oliver-sanders** [comment]
The cleanest way I can think to do that would be:

```
cylc reg [NAME] [PATH]
```

With the following being equivalent:

```
cylc reg
cylc reg $(basename $PWD)
cylc reg $(basename $PWD) $PWD
```

**Comment by hjoliver** [comment]
Have not address the "no funny chars", due to this comment:  https://github.com/cylc/cylc/issues/2281#issuecomment-360176816

**Comment by hjoliver** [comment]
@oliver-sanders - implemented as you suggested (which turned out to be mildly fiendish due to subtly different command line semantics for `cylc register` vs on-the-fly registration by `cylc run`).

**Comment by oliver-sanders** [comment]
Sorry it's taken a while to come back to this. All looks good and tests as expected, I'm pretty happy with this new functionality.

My only comment is that with the new `cylc run` "run and register" system we are unable to see what suite is being run just be inspecting the command being run:

```console
$ ps -fu osanders | grep cylc
user 1234     1  2 12:34 ?        00:00:00 python2 /path/to/cylc/cylc-run foo  # cylc run foo
user 1234     1  2 12:34 ?        00:00:00 python2 /path/to/cylc/cylc-run      # cylc run
```

This will mess up our suite monitoring unless we can come up with a better way of doing things. I guess we could `exec` the `cylc run` with the "full" arguments but is is worth it?

**Comment by hjoliver** [comment]
@oliver-sanders - that is unfortunate (what do you mean by "our suite monitoring" exactly?). Do we need to disable auto-registration for that reason?

**Comment by hjoliver** [comment]
I suppose your `exec` idea could work if I moved auto-registration out of `scheduler.py` and into the main `cylc run` command.

**Comment by matthewrmshin** [comment]
(We have tools to monitor user suites running on our dedicated suite hosts. The tools all use `ps` to list suite server program processes for a list of what users are running what.)

**Comment by hjoliver** [comment]
(We don't have such tools - that I'm aware of - but I certainly use `ps` myself a lot for the same purpose.)

**Comment by oliver-sanders** [comment]
If it doesn't sacrifice design it would be good to be able to see what suite is being run from the command line, not just for the monitoring but general usage and system maintenance.

**Comment by hjoliver** [comment]
@oliver-sanders @matthewrmshin - ready for re-review.  See updated description at top of page.

File: lib/cylc/scheduler_cli.py [hunk 3]
```diff
@@ -89,6 +90,13 @@
 def main(is_restart=False):
     """CLI main."""
     options, args = parse_commandline(is_restart)
+    if not args:
+        # Auto-registration: "cylc run" (no args) in source dir.
+        reg = SuiteSrvFilesManager().register(on_the_fly=True)
+        pth, exc = os.path.split(sys.argv[0])
```
Comment (oliver-sanders) [inline lib/cylc/scheduler_cli.py]: Unused vars?

File: bin/cylc-register [hunk 4]
```diff
@@ -18,30 +18,36 @@
...
+sub-directory containing authentication files and a "source" symlink to PATH.
 
-Suite names are the same as the directory path under the suite run directory.
-They may contain alphanumeric characters plus '_' '-' and '/'.
+Suite names can be hierarchical, corresponding to the path under ~/cylc-run.
 
-Example: if the cylc run directory is $HOME/cylc-run (the default) and
-/home/bob/suites/test is a suite source directory, then:
+  % cylc register dogs/fido PATH
+Register PATH/suite.rc as dogs/fido, with run directory ~/cylc-run/dogs/fido.
```
Comment (matthewrmshin) [inline bin/cylc-register]: A suite for dogs! :dog: :dog2:

**Comment by matthewrmshin** [comment]
(Running test battery in my environment. I'll merge on completion.)

**Comment by matthewrmshin** [comment]
Unable to get these to pass at the moment:

```
./tests/authentication/10-remote-suite-passphrase-cache.t (Wstat: 0 Tests: 5 Failed: 5)
  Failed tests:  1-5
./tests/authentication/08-shared-fs.t                    (Wstat: 0 Tests: 4 Failed: 4)
  Failed tests:  1-4
```

Tests seem to have problem locating suite source directory on remote host. E.g. tests/authentication/10 is giving me this:

```
ERROR: no suite.rc in /home/matt
ERROR: command terminated by signal 1: ssh -oBatchMode=yes -oConnectTimeout=10 myhost env CYLC_VERSION=7.7.1-173-gf3cf5 bash --login -c ''"'"'exec "$0" "$@"'"'"'' '$HOME/cylc.git/bin/cylc' register cylctb-20181015T104653Z/authentication/10-remote-suite-passphrase-cache
scp: cylc-run/cylctb-20181015T104653Z/authentication/10-remote-suite-passphrase-cache: No such file or directory
```

**Comment by hjoliver** [comment]
@matthewrmshin - those two tests now pass here, with a remote test host configured. Hopefully good to go now.

File: lib/cylc/scheduler_cli.py [hunk 5]
```diff
@@ -72,6 +83,11 @@
 def main(is_restart=False):
     """CLI main."""
     options, args = parse_commandline(is_restart)
+    if not args:
+        # Auto-registration: "cylc run" (no args) in source dir.
+        reg = SuiteSrvFilesManager().register()
+        # Replace this process with "cylc run REG ..." for easy identification.
+        os.execv(sys.argv[0], [sys.argv[0]] + [reg] + sys.argv[1:])
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply lib/cylc/scheduler_cli.py]: Anyone know if we need to /can do anything about this?

[the remark this answers is not in the document]
Comment (matthewrmshin) [reply lib/cylc/scheduler_cli.py]: I believe this is actually the cleanest way to implement what is desired.

[the remark this answers is not in the document]
Comment (kinow) [reply lib/cylc/scheduler_cli.py]: Indeed. I had a look and couldn't find a way to exploit it... unless a user managed to inject something in sys.argv[1:], but I think we have enough validation for now 👍

[the remark this answers is not in the document]
Comment (hjoliver) [reply lib/cylc/scheduler_cli.py]: OK, I'll ignore codacy on this, and punt it to the yet-to-be-fully-decided issue of how to "address" codacy-flagged potential issues that aren't real issues.

**Comment by hjoliver** [comment]
@matthewrmshin - I think you can merge this after confirming the two tests above work in your environment.

**Comment by matthewrmshin** [comment]
`tests/authentication/10-remote-suite-passphrase-cache.t` is working, but `./tests/authentication/08-shared-fs.t` is hanging for some reason. I'll investigate.

**Comment by matthewrmshin** [comment]
Turns out it needs something like this:

```diff
diff --git a/tests/authentication/08-shared-fs.t b/tests/authentication/08-shared-fs.t
index da2bfc8b7..27632135f 100755
--- a/tests/authentication/08-shared-fs.t
+++ b/tests/authentication/08-shared-fs.t
@@ -26,7 +26,7 @@ export CYLC_TEST_HOST=$( \
 if [[ -z "${CYLC_TEST_HOST}" ]]; then
     skip_all '"[test battery]remote host with shared fs": not defined'
 fi
-set_test_number 4
+set_test_number 5
 
 # "install_suite" does not work here because it installs suites on the TMPDIR,
 # which is often on local file systems. We need to ensure that the suite
@@ -35,9 +35,9 @@ SUITE_NAME="cylctb-${CYLC_TEST_TIME_INIT}/${TEST_SOURCE_DIR_BASE}/${TEST_NAME_BA
 SUITE_RUN_DIR="$(cylc get-global-config --print-run-dir)/${SUITE_NAME}"
 mkdir -p "$(dirname "${SUITE_RUN_DIR}")"
 cp -r "${TEST_SOURCE_DIR}/${TEST_NAME_BASE}" "${SUITE_RUN_DIR}"
-cylc register "${SUITE_NAME}" "${SUITE_RUN_DIR}" 2>'/dev/null'
+run_ok "${TEST_NAME_BASE}-register" cylc register --host="${CYLC_TEST_HOST}" "${SUITE_NAME}" "${SUITE_RUN_DIR}"
 
-run_ok "${TEST_NAME_BASE}-validate" cylc validate "${SUITE_NAME}"
+run_ok "${TEST_NAME_BASE}-validate" cylc validate "${SUITE_RUN_DIR}"
 
 cylc run --debug --no-detach --reference-test --host="${CYLC_TEST_HOST}" "${SUITE_NAME}" \
     1>'out' 2>&1 &
```

There may be some wider problem here, however:
1. Doing the registration on originator host creates a SSL certification problem. (I don't think we had this problem before?)
2. Not doing the registration at all causes `cylc run ...` to fail.

**Comment by hjoliver** [comment]
(sorry, let me check that one again, I think I forgot to add a final change to the branch)

**Comment by hjoliver** [comment]
(OK I think I hadn't tested 08 with a real shared fs host; it's hanging for me too).

**Comment by hjoliver** [comment]
It turns out on master the suite daemon (re)registers every suite again at start-up (on the suite host, obviously), which I think regenerates the SSL (on the suite host):

https://github.com/cylc/cylc/blob/master/lib/cylc/scheduler.py#L125-L128

On this branch I don't do that, hence why you had to invoke `cylc register` on the suite host to get the test to work.  Unfortunately I don't recall why we were doing a re-register on master.  It doesn't seem like the right thing to do. If it is, we should make the register command only create directories and the source symlink, and also have the ssl cert created by the daemon at start-up.  Any thoughts?  (I need to quit for the night now...)

**Comment by matthewrmshin** [comment]
I think it should only regenerate the SSL certificate if the current host name differs. However the behaviour is also dependent on the version of the SSL library. Older version does not have the functionality to extract the information from the old certificate.

**Comment by matthewrmshin** [comment]
(So maybe it ends up generating a new certificate every run because it is not able to do the host match?)

**Comment by matthewrmshin** [comment]
(Sorry for making things complicated!)

**Comment by hjoliver** [comment]
(No worries, I think I know the way forward).

**Comment by hjoliver** [comment]
Last fix separates run director and auth files creation, and renews the ssl.cert on the suite host at suite startup.  So this preserves current behaviour (with respect to ssl.cert) on master.  Test should pass now.

**Comment by matthewrmshin** [comment]
All good in my environment.

**Comment by matthewrmshin** [comment]
All good in my environment.

**Comment by matthewrmshin** [comment]
@oliver-sanders please do a final sanity check.

**Comment by oliver-sanders** [comment]
Sorry!

```console
$ cylc run elephant
Traceback (most recent call last):
  File "cylc/bin/cylc-run", line 25, in <module>
    main(is_restart=False)
  File "cylc/lib/cylc/scheduler_cli.py", line 116, in main
    scheduler = Scheduler(is_restart, options, args)
  File "cylc/lib/cylc/scheduler.py", line 126, in __init__
    self.suite)
  File "cylc/lib/cylc/suite_srv_files_mgr.py", line 327, in get_suite_source_dir
    raise SuiteServiceFileError("ERROR: Suite not found %s" % reg)
cylc.suite_srv_files_mgr.SuiteServiceFileError: ERROR: Suite not found elephant
```

**Comment by hjoliver** [comment]
Ah ... do you have a suite named 'elephant'?  I guess your point is, for consistency, it shouldn't drop a traceback without `--debug`?

**Comment by oliver-sanders** [comment]
Looks good!

**Pull request merged**